### PR TITLE
Bump LLVM to 84fe34a0b7fd

### PR DESCRIPTION
**Pull request by hanchenye**
Quite some changes are made... Several things I'm not sure about:
1. FunctionLike is updated from a trait to FunctionOpInterface, please take a look at the .td files whether the updates are reasonable :)
2. We need to explicitly specify the `Location`s when adding arguments to a block. For those obvious cases (such as one-to-one conversion), I use the location of the original argument; For the other cases, I'm using UnknownLocation for now. Please feel free to commit on this branch or open other PRs to improve them.

The only test case that cannot pass is Dialect/FIRRTL/canonicalization.mlir, because we no longer have `getAsmResultNames` for `OpAsmDialectInterface`, see https://reviews.llvm.org/D117750. This means the current FIRRTL dialect's `getAsmResultNames` cannot work now. I replaced that method with multiple `getAsmResultNames`s owned by concrete ops (including Instance/MemOp, Node/Reg/RegReset/WireOp, Constant/InvalidValueOp, and several CIRRTL ops). However, I haven't find a way to handle the `name` attribute, which causes the failed test case:
```cpp
    // Many firrtl dialect operations have an optional 'name' attribute.  If
    // present, use it.
    if (op->getNumResults() == 1)
      if (auto nameAttr = op->getAttrOfType<StringAttr>("name"))
        setNameFn(op->getResult(0), nameAttr.getValue());
```
FileCheck result:
```shell
/circt/test/Dialect/FIRRTL/canonicalization.mlir:1814:17: error: CHECK-NEXT: expected string not found in input
 // CHECK-NEXT: %bar0 = firrtl.bits %0 4 to 0 {name = "bar0"} : (!firrtl.uint<6>) -> !firrtl.uint<5>
                ^
<stdin>:566:82: note: scanning from here
 %0 = firrtl.add %d1, %d1 : (!firrtl.uint<5>, !firrtl.uint<5>) -> !firrtl.uint<6>
                                                                                 ^
<stdin>:567:2: note: possible intended match here
 %1 = firrtl.bits %0 4 to 0 {name = "bar0"} : (!firrtl.uint<6>) -> !firrtl.uint<5>
 ^
```

### CODE REVIEW

File: include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td [hunk 1]
```diff
@@ -10,8 +10,11 @@
 //
 //===----------------------------------------------------------------------===//
 
+def HasCustomSSAName : DeclareOpInterfaceMethods<OpAsmOpInterface,
```
Comment (darthscsi) [inline include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td:13]: Could this just be added to FIRRTLOp then all ops would get it?

Comment (hanchenye) [reply include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td:13]: Does that also mean each FIRRTL op will have to have its own getAsmResultNames? Then, there will be a large duplication in FIRRTLOps.cpp...

**Review by teqdruid — APPROVED** [review]
Changes to the MSFT dialect look acceptable to me! Thanks

**Comment by darthscsi** [comment]
I've a small llvm bump to right before the OpAsm stuff to make this PR smaller here: https://github.com/llvm/circt/pull/2577

**Comment by hanchenye** [comment]
> I've a small llvm bump to right before the OpAsm stuff to make this PR smaller here: #2577

Nice, thanks! I've rebased the branch.

**Comment by darthscsi** [comment]
I've folded this into here: https://github.com/llvm/circt/pull/2638 since it was getting old.

**Pull request closed without merging**